nexus/db-queries/src/db/datastore/reconfigurator_chicken_switches.rs

```diff
@@ -0,0 +1,332 @@
+    }
+
+    pub async fn reconfigurator_chicken_switches_get(
+        &self,
+        opctx: &OpContext,
+        version: u32,
+    ) -> Result<Option<ReconfiguratorChickenSwitches>, Error> {
+        opctx.authorize(authz::Action::Read, &authz::BLUEPRINT_CONFIG).await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+        use nexus_db_schema::schema::reconfigurator_chicken_switches::dsl;
+
+        let latest = dsl::reconfigurator_chicken_switches
+            .filter(dsl::version.eq(SqlU32::new(version)))
+            .select(DbReconfiguratorChickenSwitches::as_select())
+            .get_result_async(&*conn)
+            .await
+            .optional()
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(latest.map(Into::into))
+    }
+
+    /// Insert the current version of the chicken switches in the database
+    ///
+    /// Only succeeds if the prior version is the latest version currently in
+    /// the `reconfigurator_chicken_switches` table. If there are no versions
+    /// currently in the table, then the current swtiches must be at version 1.
+    pub async fn reconfigurator_chicken_switches_insert_latest_version(
+        &self,
+        opctx: &OpContext,
+        switches: ReconfiguratorChickenSwitchesParam,
+    ) -> Result<(), Error> {
+        if switches.version < 1 {
+            return Err(Error::invalid_request(
+                "version must be greater than 0",
+            ));
+        }
+
+        let ReconfiguratorChickenSwitchesParam { version, planner_enabled } =
+            switches;
+        let switches = ReconfiguratorChickenSwitches {
+            version,
+            planner_enabled,
+            time_modified: chrono::Utc::now(),
+        };
+
+        opctx
+            .authorize(authz::Action::Modify, &authz::BLUEPRINT_CONFIG)
+            .await?;
+
+        let num_inserted = self
+            .reconfigurator_chicken_switches_insert_next_version(
+                opctx, &switches,
+            )
+            .await?;
+
+        match num_inserted {
+            0 => Err(Error::invalid_request(format!(
+                "version {} is not the most recent",
+                switches.version
+            ))),
+            1 => Ok(()),
+            // This is impossible because we are explicitly inserting only one
+            // row with a unique primary key.
+            _ => unreachable!("query inserted more than one row"),
+        }
+    }
+
+    /// Insert the next version of the chicken switches in the database
+    ///
+    /// Only succeeds if the prior version is the latest version currently
+    /// in the `reconfigurator_chicken_switches` table.
+    ///
+    /// Panics if `switches.version < 1`;
+    async fn reconfigurator_chicken_switches_insert_next_version(
+        &self,
+        opctx: &OpContext,
+        switches: &ReconfiguratorChickenSwitches,
+    ) -> Result<usize, Error> {
+        assert!(switches.version >= 1);
```

Comment:
Ok, you've convinced me that `version` needs to be supplied by the client. But that means this needs to return a `bad_request` error instead of panicking, I think? I realize omdb will never be evil here, but we shouldn't panic if we get a request with version=0.